### ISSUE TITLE
lint_ev_organization_id_missing incorrectly flags CA certificates without cabfOrganizationIdentifier

### ISSUE THREAD

**Issue by om226** [zmap/zlint#1005]
**Summary**  
The lint `lint_ev_organization_id_missing` (which sets error `e_ev_organization_id_missing`) is being raised on EV CA certificates that do *not* include the `cabfOrganizationIdentifier` extension, even though per the CAB Forum Extended Validation Guidelines, that extension is *not required* for CA certificates unless the subject has an `organizationIdentifier`.

**Details**  
- In the CAB Forum EV Guidelines v2.0.1, section 7.1.2.2 (“CA/Browser Forum Organization Identifier Extension”), the `cabfOrganizationIdentifier` extension is marked as **Optional**. :contentReference[oaicite:7]{index=7}  
- The guideline states: *“If the subject:organizationIdentifier is present, this field MUST be present.”* :contentReference[oaicite:8]{index=8}  
- There is **no requirement** in the EV Guidelines that *all* CA certificates must have `cabfOrganizationIdentifier` — only when the subject has an `organizationIdentifier`. Also, the guidelines explicitly note that no extension is mandatory on a CA unless it is stated as “Required” in the relevant subsection. :contentReference[oaicite:9]{index=9}  
- I have identified at least two EV CA certificates where `zlint` currently reports `e_ev_organization_id_missing`, despite the absence of `organizationIdentifier` in the subject:
  - Certificate 1: https://crt.sh/?id=10305004920&opt=pkimetal
  - Certificate 2: https://crt.sh/?id=21408224858&opt=pkimetal
- In both cases, the behavior seems to be a **false positive**.

**Impact**  
- This lint produces incorrect error reports for legitimate CA certificates, which can lead to confusion, misdiagnosis, or unnecessary CA configuration changes.  
- Users of *zlint* depending on strict EV compliance checks may be incorrectly blocked or misinformed.

**Proposed fix / change**  
- Update `lint_ev_organization_id_missing` to **skip** CA certificates (i.e., certificates with `cA=true` in BASIC_CONSTRAINTS) when the subject does *not* have `organizationIdentifier`. 

**References**  
- CAB Forum EV Guidelines v2.0.1, section 7.1.2.2.

**Comment by digirenpeter** [comment]
Both certificates mentioned in the report do appear to contain a subject:organizationIdentifier:
```
organizationIdentifier = VATDE-202620438
organizationIdentifier = VATRO-18288250
```

The current lint in question only applies when len(c.Subject.OrganizationIDs) > 0:
```
func (l *evOrgIdExtMissing) CheckApplies(c *x509.Certificate) bool {
    return util.IsEV(c.PolicyIdentifiers) &&
        len(c.Subject.OrganizationIDs) > 0
}
```

So the lint is not enforcing cabfOrganizationIdentifier for all EV CA certificates, but only for EV certificates that include subject:organizationIdentifier, which matches EVG 7.1.2.2:
> Extension Name: cabfOrganizationIdentifier (OID: 2.23.140.3.1)
> Verbose OID: {joint-iso-itu-t(2) international-organizations(23)
> ca-browser-forum(140) certificate-extensions(3)
> cabf-organization-identifier(1) }
> Required/Optional: Optional (but see below)
> Contents: If the subject:organizationIdentifier is present, this field MUST be present.

Based on the subjects above, the lint appears to be behaving as intended. Unless I’m missing something, these examples seem to satisfy the exact condition that the lint is checking for.

**Comment by om226** [comment]
I'd like to clarify the original issue and provide additional evidence.

Both CA certificates referenced in this issue contain `subject:organizationIdentifier`, but neither contains the `cabfOrganizationIdentifier` extension.

ZLint v3.7.1 reports `e_ev_organization_id_missing` for both certificates, as shown in the attached screenshots.

The actual issue is that **the lint enforces this requirement on CA certificates, not just End-Entity certificates.**

According to the CA/Browser Forum EV Guidelines v2.0.4:

* **Section 7.1.2** explicitly states that no extension or extension format is mandatory on a CA unless specifically marked as "Required" in the relevant subsection.
* **Section 7.1.2.2** marks `cabfOrganizationIdentifier` as "Optional (but see below)" and introduces a conditional MUST requirement when `subject:organizationIdentifier` is present.

My interpretation is that the conditional requirement in §7.1.2.2 does not override the explicit CA exemption in §7.1.2.

The current implementation of `CheckApplies()` does not distinguish between CA and End-Entity certificates:

```go
return util.IsEV(c.PolicyIdentifiers) &&
    len(c.Subject.OrganizationIDs) > 0
```

**Proposed change:**

```go
return !c.IsCA &&
    util.IsEV(c.PolicyIdentifiers) &&
    len(c.Subject.OrganizationIDs) > 0
```

This would exclude CA certificates from this lint while preserving the existing validation for EV End-Entity certificates.

The original issue description should also be corrected: both example certificates contain `subject:organizationIdentifier`. The proposed fix should therefore exclude CA certificates regardless of whether this attribute is present.

Could you please confirm whether this interpretation of the EV Guidelines is correct and whether the lint should be restricted to End-Entity certificates?

**Comment by digirenpeter** [comment]
Someone else can chime in but my interpretation isn't that section 7.1.2 provides a blanket exemption for CA certificates.
> 7.1.2 Certificate extensions
> The extensions listed in Section 7.1.2 are recommended for maximum interoperability
> between certificates and browsers / applications, but are not mandatory on the CAs
> except where indicated as “Required”.

and:

> If a CA includes an extension in a certificate that has a Certificate field which is named
> in Section 7.1.2, the CA must follow the format specified in that subsection. However,
> no extension or extension format shall be mandatory on a CA unless specifically stated
> as “Required” in the subsection that describes the extension.

Section 7.1.2 states that extensions are not mandatory for CAs unless the subsection describing the extension specifically requires them. Section 7.1.2.2 contains such a conditional requirement:
> Extension Name: cabfOrganizationIdentifier (OID: 2.23.140.3.1)
> Verbose OID: {joint-iso-itu-t(2) international-organizations(23) ca-
> browser-forum(140) certificate-extensions(3) cabf-organization-
> identifier(1) }
> Required/Optional: Optional (but see below)
> Contents: If the subject:organizationIdentifier is present, this field MUST be present.

The "Optional (but see below)" wording indicates that the extension is optional unless the stated condition applies. Once subject:organizationIdentifier is present, the conditional MUST requirement applies.

So an EV CA certificate containing subject:organizationIdentifier would also need to contain cabfOrganizationIdentifier.

I don't believe the "Optional" designation alone creates an exemption for CA certificates, because the subsection explicitly defines a condition under which the extension becomes mandatory.

**Comment by om226** [comment]
Thanks for your feedback. I found two relevant Mozilla CA compliance investigations that directly address this interpretation:

1. DigiCert – https://bugzilla.mozilla.org/show_bug.cgi?id=2056663
2. D-Trust – https://bugzilla.mozilla.org/show_bug.cgi?id=2056882

In the DigiCert investigation, the CA cites a formal response provided by the CA/Browser Forum Chair on February 9, 2026, following discussions among CABF members. According to that response, the extension requirements in EV Guidelines §7.1.2, including cabfOrganizationIdentifier, apply only to Subscriber Certificates. The requirements for extensions in EV CA certificates are addressed in §7.1.4.3.

The D-Trust investigation concerns one of the exact certificates referenced in this issue. D-Trust confirmed that the certificate contains subject:organizationIdentifier but lacks cabfOrganizationIdentifier. After reviewing the applicable requirements and referencing the CABF clarification, D-Trust concluded that the certificate was not misissued.

Both Mozilla incidents were closed as RESOLVED INVALID.

This provides additional evidence that the lint should exclude CA certificates, regardless of whether subject:organizationIdentifier is present.

The proposed CheckApplies() change remains:

```go
return !c.IsCA &&
    util.IsEV(c.PolicyIdentifiers) &&
    len(c.Subject.OrganizationIDs) > 0
```

This preserves the conditional requirement for EV Subscriber certificates while excluding CA certificates from a requirement that the CABF clarification identifies as subscriber-specific.

I also want to correct the original issue description: both example CA certificates contain subject:organizationIdentifier. The issue is the applicability of the extension requirement to CA certificates, not the absence of the subject attribute.

Given these findings, would you reconsider the proposed change?

**Comment by digirenpeter** [comment]
I think the discussion on those bugzilla tickets clears up the ambiguity so I've opened this PR with the recommended fix:
https://github.com/zmap/zlint/pull/1091

**Comment by om226** [comment]
Thank you for looking into this and for opening the PR! I really appreciate your help in getting this resolved.
